File: src/backend/folha/src/model/folha_lancamento_cabecalho_model.py
from src import db
from src.model.folha_lancamento_detalhe_model import FolhaLancamentoDetalheModel
from src.model.view_pessoa_colaborador_model import ViewPessoaColaboradorModel


class FolhaLancamentoCabecalhoModel(db.Model):
    __tablename__ = 'folha_lancamento_cabecalho'

    id = db.Column(db.Integer, primary_key=True)
    competencia = db.Column(db.String(7))
    tipo = db.Column(db.String(1))
    id_colaborador = db.Column(db.Integer, db.ForeignKey('view_pessoa_colaborador.id'))

    folha_lancamento_detalhe_model_list = db.relationship('FolhaLancamentoDetalheModel', lazy='dynamic')
    view_pessoa_colaborador_model = db.relationship('ViewPessoaColaboradorModel', foreign_keys=[id_colaborador])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_colaborador = data.get('idColaborador')
        self.competencia = data.get('competencia')
        self.tipo = data.get('tipo')

    def serialize(self):
        return {
            'id': self.id,
            'idColaborador': self.id_colaborador,
            'competencia': self.competencia,
            'tipo': self.tipo,
            'folhaLancamentoDetalheModelList': [folha_lancamento_detalhe_model.serialize() for folha_lancamento_detalhe_model in self.folha_lancamento_detalhe_model_list],
            'viewPessoaColaboradorModel': self.view_pessoa_colaborador_model.serialize() if self.view_pessoa_colaborador_model else None,
        }

File: frontend/folha/lib/app/data/model/folha_lancamento_cabecalho_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:folha/app/data/model/model_imports.dart';
import 'package:folha/app/data/domain/domain_imports.dart';

class FolhaLancamentoCabecalhoModel {
	int? id;
	int? idColaborador;
	String? competencia;
	String? tipo;
	List<FolhaLancamentoDetalheModel>? folhaLancamentoDetalheModelList;
	ViewPessoaColaboradorModel? viewPessoaColaboradorModel;

	FolhaLancamentoCabecalhoModel({
		this.id,
		this.idColaborador,
		this.competencia,
		this.tipo,
		this.folhaLancamentoDetalheModelList,
		this.viewPessoaColaboradorModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'competencia',
		'tipo',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Competencia',
		'Tipo',
	];

	FolhaLancamentoCabecalhoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idColaborador = jsonData['idColaborador'];
		competencia = jsonData['competencia'];
		tipo = FolhaLancamentoCabecalhoDomain.getTipo(jsonData['tipo']);
		folhaLancamentoDetalheModelList = (jsonData['folhaLancamentoDetalheModelList'] as Iterable?)?.map((m) => FolhaLancamentoDetalheModel.fromJson(m)).toList() ?? [];
		viewPessoaColaboradorModel = jsonData['viewPessoaColaboradorModel'] == null ? ViewPessoaColaboradorModel() : ViewPessoaColaboradorModel.fromJson(jsonData['viewPessoaColaboradorModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idColaborador'] = idColaborador != 0 ? idColaborador : null;
		jsonData['competencia'] = competencia;
		jsonData['tipo'] = FolhaLancamentoCabecalhoDomain.setTipo(tipo);
		
		var folhaLancamentoDetalheModelLocalList = []; 
		for (FolhaLancamentoDetalheModel object in folhaLancamentoDetalheModelList ?? []) { 
			folhaLancamentoDetalheModelLocalList.add(object.toJson); 
		}
		jsonData['folhaLancamentoDetalheModelList'] = folhaLancamentoDetalheModelLocalList;
		jsonData['viewPessoaColaboradorModel'] = viewPessoaColaboradorModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idColaborador = plutoRow.cells['idColaborador']?.value;
		competencia = plutoRow.cells['competencia']?.value;
		tipo = plutoRow.cells['tipo']?.value != '' ? plutoRow.cells['tipo']?.value : 'Folha Mensal';
		folhaLancamentoDetalheModelList = [];
		viewPessoaColaboradorModel = ViewPessoaColaboradorModel();
		viewPessoaColaboradorModel?.nome = plutoRow.cells['viewPessoaColaboradorModel']?.value;
	}	

	FolhaLancamentoCabecalhoModel clone() {
		return FolhaLancamentoCabecalhoModel(
			id: id,
			idColaborador: idColaborador,
			competencia: competencia,
			tipo: tipo,
			folhaLancamentoDetalheModelList: folhaLancamentoDetalheModelListClone(folhaLancamentoDetalheModelList!),
		);			
	}

	folhaLancamentoDetalheModelListClone(List<FolhaLancamentoDetalheModel> folhaLancamentoDetalheModelList) { 
		List<FolhaLancamentoDetalheModel> resultList = [];
		for (var folhaLancamentoDetalheModel in folhaLancamentoDetalheModelList) {
			resultList.add(
				FolhaLancamentoDetalheModel(
					id: folhaLancamentoDetalheModel.id,
					idFolhaLancamentoCabecalho: folhaLancamentoDetalheModel.idFolhaLancamentoCabecalho,
					idFolhaEvento: folhaLancamentoDetalheModel.idFolhaEvento,
					origem: folhaLancamentoDetalheModel.origem,
					provento: folhaLancamentoDetalheModel.provento,
					desconto: folhaLancamentoDetalheModel.desconto,
				)
			);
		}
		return resultList;
	}

	
}